// jeep-sqlite/dist/collection/utils/utils-store.js
export const getDBFromStore = async (dbName, store) => {
  try {
    const retDb = await store.getItem(dbName);
    return Promise.resolve(retDb);
  }
  catch (err) {
    return Promise.reject(`GetDBFromStore: ${err.message}`);
  }
};
export const setInitialDBToStore = async (dbName, store) => {
  try {
    // export the database
    const data = null;
    // store the database
    await store.setItem(dbName, data);
    return Promise.resolve();
  }
  catch (err) {
    return Promise.reject(`SetInitialDBToStore: ${err.message}`);
  }
};
export const setDBToStore = async (mDb, dbName, store) => {
  try {
    // export the database
    const data = mDb.export();
    // store the database
    await saveDBToStore(dbName, data, store);
    return Promise.resolve();
  }
  catch (err) {
    return Promise.reject(`SetDBToStore: ${err.message}`);
  }
};
export const saveDBToStore = async (dbName, data, store) => {
  try {
    // store the database
    await store.setItem(dbName, data);
    return Promise.resolve();
  }
  catch (err) {
    return Promise.reject(`SaveDBToStore: ${err.message}`);
  }
};
export const removeDBFromStore = async (dbName, store) => {
  try {
    await store.removeItem(dbName);
    return Promise.resolve();
  }
  catch (err) {
    return Promise.reject(`RemoveDBFromStore: ${err.message}`);
  }
};
export const isDBInStore = async (dbName, store) => {
  try {
    const retDb = await store.getItem(dbName);
    if (retDb != null && retDb.length > 0) {
      return Promise.resolve(true);
    }
    else {
      return Promise.resolve(false);
    }
  }
  catch (err) {
    return Promise.reject(`IsDBInStore: ${err}`);
  }
};
export const restoreDBFromStore = async (dbName, prefix, store) => {
  const mFileName = `${prefix}-${dbName}`;
  try {
    // check if file exists
    const isFilePre = await isDBInStore(mFileName, store);
    if (isFilePre) {
      const isFile = await isDBInStore(dbName, store);
      if (isFile) {
        const retDb = await getDBFromStore(mFileName, store);
        await saveDBToStore(dbName, retDb, store);
        await removeDBFromStore(mFileName, store);
        return Promise.resolve();
      }
      else {
        return Promise.reject(new Error(`RestoreDBFromStore: ${dbName} does not exist`));
      }
    }
    else {
      return Promise.reject(new Error(`RestoreDBFromStore: ${mFileName} does not exist`));
    }
  }
  catch (err) {
    return Promise.reject(`RestoreDBFromStore: ${err.message}`);
  }
};
export const copyDBToStore = async (dbName, toDb, store) => {
  try {
    // check if file exists
    const isFile = await isDBInStore(dbName, store);
    if (isFile) {
      const retDb = await getDBFromStore(dbName, store);
      await saveDBToStore(toDb, retDb, store);
      return Promise.resolve();
    }
    else {
      return Promise.reject(new Error(`CopyDBToStore: ${dbName} does not exist`));
    }
  }
  catch (err) {
    return Promise.reject(`CopyDBToStore: ${err.message}`);
  }
};
export const getDBListFromStore = async (store) => {
  try {
    const retDbList = await store.keys();
    return Promise.resolve(retDbList);
  }
  catch (err) {
    return Promise.reject(`GetDBListFromStore: ${err.message}`);
  }
};
